### PR TITLE
[WIP] ERL core logic

### PR DESCRIPTION
**Pull request by ecita**

### CODE REVIEW

File: lib/db.js [hunk 1]
```diff
@@ -239,39 +246,77 @@ class LimitDBRedis extends EventEmitter {
...
+      }
+      this.redis.takeElevated(key, erlIsActiveKey,
+          bucketKeyConfig.ms_per_interval || 0,
+          bucketKeyConfig.size,
+          count,
+          Math.ceil(bucketKeyConfig.ttl || this.globalTTL),
+          bucketKeyConfig.drip_interval || 0,
+          bucket.elevated_limits.ms_per_interval || 0,
+          bucket.elevated_limits.size,
+          bucketKeyConfig.elevated_limits.activation_period_minutes || 15,
```
Comment (panga) [inline lib/db.js:263]: Please include all new client & bucket configurations in the README.

File: lib/take_elevated.lua [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    end
+
+    enough_tokens = new_content >= tokens_to_take
+    if enough_tokens then
+        new_content = math.min(new_content - tokens_to_take, erl_bucket_size)
+    end
+end
+
+-- https://redis.io/commands/EVAL#replicating-commands-instead-of-scripts
+redis.replicate_commands()
```
Comment (LeweyM) [inline lib/take_elevated.lua:65]: we will need to call this before `redis.call('SET')`

File: test/db.tests.js [hunk 3]
```diff
@@ -673,6 +672,194 @@ describe('LimitDBRedis', () => {
...
+      })
+      it('should use default ttl if erl activation period is not configured',  async (done) => {
+        const bucketName = 'test-bucket';
+        await db.configurateBucket(bucketName, {
+          size: 2,
+          per_minute: 1,
+          elevated_limits: {
+            size: 10,
+            per_minute: 1,
+            activation_period_minutes: 15,
```
Comment (LeweyM) [inline test/db.tests.js:849]: this test should not have `activation_period_minutes` defined.

Also, we need another test checking that when it is defined, the defined ttl is applied.

File: lib/db.js (hunk 1 above)
Comment (LeweyM) [inline lib/db.js:263]: we can use a config variable for default `activation_period_minutes`

Comment (ecita) [reply lib/db.js:263]: Do you mean to set the default value while resolving the bucketKeyConfig itself?

File: lib/db.js [hunk 4]
```diff
@@ -340,22 +385,22 @@ class LimitDBRedis extends EventEmitter {
 
     const key = `${params.type}:${params.key}`;
     this.redis.put(key,
-      count,
```
[the remark this answers is not in the document]
Comment (ecita) [reply lib/db.js:343]: Absolutely

File: lib/take_elevated.lua (hunk 2 above)
Comment (ecita) [reply lib/take_elevated.lua:65]: Because?

File: test/db.tests.js (hunk 3 above)
Comment (ecita) [reply test/db.tests.js:849]: Yes - doing so made it obvious that we don't yet resolve the `activation_period_minutes` correctly

**Comment by ecita** [comment]
I'm closing this PR as I intended to create one merging to the fork - please continue disussion here: https://github.com/ecita/limitd-redis/pull/1

**Comment by ecita** [comment]
I'm closing this PR as I intended to create one merging to the fork - please continue disussion here: https://github.com/ecita/limitd-redis/pull/1

File: lib/take_elevated.lua (hunk 2 above)
Comment (LeweyM) [reply lib/take_elevated.lua:65]: I think Redis needs to propagate any changes to replicas before performing any writes. Not sure exactly though, and it depends on the version of Redis used.

https://redis.io/docs/interact/programmability/eval-intro/#:~:text=scripts%20debugger.-,Script%20replication,-In%20standalone%20deployments

Comment (LeweyM) [reply lib/take_elevated.lua:65]: the link to the redis doc in the code is out of date, we should replace it also

Comment (ecita) [reply lib/take_elevated.lua:65]: Addressed in new pr

**Pull request closed without merging**